src/adapter/src/coord.rs

```diff
@@ -3332,6 +3331,110 @@ impl Coordinator {
         ]);
         Ok(serde_json::Value::Object(map))
     }
+
+    /// Checks that all clusters/collections are hydrated. If so, this will
+    /// trigger `self.clusters_hydrated_trigger`.
+    ///
+    /// This method is a no-op when the trigger has already been fired.
+    async fn maybe_check_hydration_status(&mut self) {
+        if self.clusters_hydrated_trigger.is_some() {
+            let enable_caught_up_check =
+                ENABLE_0DT_CAUGHT_UP_CHECK.get(self.catalog().system_config().dyncfgs());
+
+            if enable_caught_up_check {
+                // WIP: Not sure I like this one. Is there a smarter way of
+                // finding the collection.
+                let replica_frontier_collection_id = {
+                    let entry = self
+                        .catalog()
+                        .entries()
+                        .find(|entry| entry.name().item == "mz_cluster_replica_frontiers")
+                        .expect("missing mz_cluster_replica_frontiers");
+
+                    entry.id
+                };
+
+                let live_frontiers = self
+                    .controller
+                    .storage
+                    .snapshot_latest(replica_frontier_collection_id)
+                    .await
+                    .expect("can't read mz_cluster_replica_frontiers");
+
+                let live_frontiers = live_frontiers
+                    .into_iter()
+                    .map(|row| {
+                        let mut iter = row.into_iter();
+
+                        let id = iter
+                            .next()
+                            .expect("missing object id")
+                            .unwrap_str()
+                            .to_string();
+                        let replica_id = iter
+                            .next()
+                            .expect("missing replica id")
+                            .unwrap_str()
+                            .to_string();
+                        let upper_ts = iter.next().expect("missing upper ts").unwrap_mz_timestamp();
```

Comment:
There's a known bug here. The upper can be empty and I need to handle that case. I need to get around to typing this up and will let you know.